content/concepts/hole-punching.md

```diff
@@ -0,0 +1,193 @@
+---
+title: Hole Punching
+weight: 3
+---
+
+## Types of nodes
+
+Nodes on the public internet can be divided into two groups: 
+public and non-public. Most nodes are not publicly accessible as 
+routers act as firewalls and allow for NATs. 
+
+To be clear, a non-public node from a standard LAN:
+
+- can dial other public nodes;
+- cannot dial non-public nodes.
+
+Likewise, a public node on the internet:
+
+- can dial other public nodes;
+- cannot dial non-public nodes.
+
+### How can a node dial a non-public node?
+
+Here are a few methods that nodes can use to dial a non-public node:
+
+- UPnP (Universal Plug and Play): A protocol that enables nodes 
+  to discover and connect by automatically opening
+  ports into a firewall.
+- Port forwarding: Manually configuring a port forward on a router.
+- Traversal Using Relay NAT (TURN): A protocol that can traverse
+  a NAT, allowing a client to obtain IP addresses and ports from 
+  relaying.
+
+### Limitations
+
+UPnP automates the process of node discovery and connectivity. Still, 
+it may not be available everywhere, and there can posses the risk
+of establishing connections with untrustworthy nodes as the protocol
+does not enforce authentication or authorization. Manually opening a
+port requires technical expertise and does not enforce 
+authentication or authorization. Using a relay node also requires
+technical expertise. Relaying adds additional latency and is resource
+intensive.
+
+### Possible solution: hole punching
+
+In the case where the other options aren't sufficient, networks can 
+use a technique called hole punching to establish connections with 
+non-public nodes.
+
+Each node connects to an unrestricted third-party server and 
+shares its external and internal address and port information.
+The server temporarily stores the node's information and relays 
+each node's information to the other. Clients can use this 
+information to establish direct connections with each other.
+
+Take two nodes, `A` and `B`, that would like the dial each other:
+
+1. The first packet of both nodes (e.g., in the case of TCP, an SYN) 
+   passes through their respective routers.
+2. The routers add a 5-tuple to their router's state table. 
+3. `Packet A` and `B` "punch holes" into their respective routers' 
+   firewalls.
+4. Both packets arrive at the opposite router.
+5. Once `A`'s packet arrives at `router B`, `router B` checks its state 
+   table and finds a 5-tuple previously added through the packet sent by 
+   node B. 
+6. The routers forward the packets through the "punched holes". 6. It 
+   then forwards the packet to `B`. The same occurs with `B`'s packet; 
+   upon arriving at `router A`, it matches a 5-tuple in `router A`'s state 
+   table and thus forwards the packet to `A`.
+  
+The following use case diagram illustrates the above process.
+
+![](https://i.imgur.com/0k2Zlj3.png)
```

Comment:
Hmm, how about we transform them into simpler diagrams that are direct? Here's an idea I sketched out (only shows packetA for the hole punching example and can of course be improved) 

![20220927_230108.jpg](https://user-images.githubusercontent.com/25497083/192678290-f70af19c-d17b-4d11-b2ac-d2965e6eca97.jpg)

cc. @mxinden

Reply:
I don't have an opinion here.

Another option would be to use mermaid. GitHub supports it. I am not sure our CMS does.

https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams

Reply:
That could work too. As these diagrams already exist for hole punching, does it make sense to include them in this PR and then create an issue to plan out what we can substitute them for? 

In general, I'd like to further enhance the concept docs with visuals and establish a guideline we can follow. As the docs seem to aim for high-level content and clean explainers, we may decide to have simple and fun graphics such as [this](https://developer.holochain.org/concepts/2_application_architecture/#agent-centricity), or perhaps more direct diagrams like the one above.

I also think it's good to have UML diagrams exist *somewhere*, as they *can* serve as a more formal representation.

Reply:
Would also add that we currently have similar sequence diagrams on some of the concept pages, circuit relay is an example. What do we think @marten-seemann @mxinden?

Reply:
No opinion on the diagram styles.

Reply:
I'd vote for diagrams as code instead of anything else. Either mermaid (not experienced with mermaid) or plantuml. The main problem that diagrams introduce is that they quickly become outdated, especially if they're not generated from code. I have never experienced a consistent and reliable process for keeping image-only diagrams up to date as code and feature changes are made.

Reply:
That might be acceptable for specs, but for explaining concepts in docs (and presentations we give at conferences), I think we’ll need something more visually pleasing than Plantumls.

Reply:
As a point of reference, this is an excellent post by Tailscale about NAT traversal: https://tailscale.com/blog/how-nat-traversal-works/
And this is a nice write up or TLS 1.3 by Cloudflare: https://blog.cloudflare.com/rfc-8446-aka-tls-1-3/

Reply:
Thanks folks. We'll be looking to get this off the ground asap. In the meantime, does it make sense to keep the uml diagrams in this PR and substitute them once we have new ones?

Reply:
> In the meantime, does it make sense to keep the uml diagrams in this PR and substitute them once we have new ones?

Yup, I think that's reasonable.

Reply:
@DannyS03 I guess a minor nit from me is: I'd rather commit the images to this repo (in a media/ folder)  and link to it's location there. Imo it's better than hosting it on imgur

Reply:
https://github.com/libp2p/docs/issues/207 <- Tracking issue to update diagrams once we have them available